File: Assignment3-MapReduce/mapreduce/C/mapreduce.c
#include "mapreduce.h"

int shuffle(struct Pair** mapped, size_t num_mappers, size_t doc_size,
            struct ShuffledStruct* shuffled, size_t num_keys, size_t buffer_size)
{
    int num_inserted_keys = 0;
    int was_this_key_inserted;
    int k;

    for (int i = 0; i < num_mappers; ++i) {
        for (int j = 0; j < doc_size; ++j) {
            was_this_key_inserted = 0;
            k = 0;

            while (k < num_inserted_keys && !was_this_key_inserted) {
                if (shuffled[k].key == mapped[i][j].key) {
                    was_this_key_inserted = 1;
                }
                else {
                    ++k;
                }
            }

            if (was_this_key_inserted) {
                shuffled[k].values[shuffled[k].size] = 1;
                shuffled[k].size += 1;
            }
            else {
                shuffled[num_inserted_keys].key = mapped[i][j].key;
                shuffled[num_inserted_keys].values[0] = 1;
                shuffled[num_inserted_keys].size = 1;
                num_inserted_keys += 1;
            }
        }
    }
    return num_inserted_keys;
}

void mapreduce(int* arr, size_t size, struct Pair* counts, size_t num_keys,
               void (*mapper)(int*, size_t, struct Pair*),
               void (*reducer)(struct ShuffledStruct, struct Pair*))
{
    size_t num_mappers = 4;
    size_t doc_size = size / num_mappers;

    struct Pair** mapped = malloc(num_mappers * sizeof(struct Pair*));

    for (int i = 0; i < num_mappers; ++i) {
        mapped[i] = malloc(doc_size * sizeof(pair));
    }

    for (int i = 0; i < num_mappers; ++i) {
        mapper(arr + (i * doc_size), doc_size, mapped[i]);
    }

    size_t buffer_size = size;

    struct ShuffledStruct* shuffled = malloc(num_keys * sizeof(shuffled_struct));

    for (int i = 0; i < num_keys; ++i) {
        shuffled[i].values = malloc(buffer_size * sizeof(int));
    }

    int num_inserted_keys = shuffle(mapped, num_mappers, doc_size,
                                    shuffled, num_keys, buffer_size);

    for (int i = 0; i < num_inserted_keys; ++i) {
        reducer(shuffled[i], &counts[i]);
    }

    for (int i = 0; i < num_keys; ++i) {
        free(shuffled[i].values);
    }

    free(shuffled);

    for (int i = 0; i < num_mappers; ++i) {
        free(mapped[i]);
    }

    free(mapped);
}

File: Assignment3-MapReduce/mapreduce/Python/mapreduce.py
"""
Simple implementation of MapReduce in Python
(c) Oleksandr Zaytsev <oleks@ucu.edu.ua>
"""

from multiprocessing import Process, Queue
import time


def split(arr, num_mappers):
    """Splits an array into a set of documents"""
    doc_size = len(arr) // num_mappers
    return [arr[i * doc_size:(i + 1) * doc_size] for i in range(num_mappers)]


def shuffle(mapped, num_reducers):
    shuffled = {}

    for pairs in mapped:
        for pair in pairs:
            k = pair[0]
            v = pair[1]

            if k in shuffled.keys():
                shuffled[k].append(v)
            else:
                shuffled[k] = [v]

    return shuffled


def mapreduce_sequential(arr, num_mappers, mapper_func, reducer_func):
    splitted = split(arr, num_mappers)
    mapped = [mapper_func(doc) for doc in splitted]
    shuffled = shuffle(mapped, None)
    reduced = dict([reducer_func(k, v) for k, v in shuffled.items()])
    return dict(sorted(reduced.items()))


def parallel_map(doc, output, mapper_func):
    mapped = mapper_func(doc)
    output.put(mapped)


def parallel_reduce(key, value, output, reducer_func):
    reduced = reducer_func(key, value)
    output.put(reduced)


def mapreduce_parallel(arr, num_cores, mapper_func, reducer_func):
    splitted = split(arr, num_cores)
    
    output = Queue()
    processes = [
        Process(target=parallel_map, args=(doc, output, mapper_func))
        for doc in splitted]

    for p in processes:
        p.start()

    for p in processes:
        p.join()

    mapped = [output.get() for p in processes]
    shuffled = shuffle(mapped, None)

    output = Queue()
    processes = [
        Process(target=parallel_reduce, args=(key, value, output, reducer_func))
        for key, value in shuffled.items()]

    for p in processes:
        p.start()

    for p in processes:
        p.join()

    return dict(sorted([output.get() for p in processes]))
